### PR TITLE
Upload coverage from selfhosted

### PR DESCRIPTION
**Pull request by stavros11**
Final implementation of #265 with the commits squashed to one to clean history. I think this should work but I still need to check how the coverage is logged in the website.

It appears that the coverage is increased by 13% since parts of the instrument code are now included. However I think the coverage of the test files is not logged properly because I am using symbolic links to put them in the /tmp/ directory.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Comment by scarrazza** [comment]
Thank you @stavros11, very good. Could you please try to clone directly inside `/tmp/`?

**Comment by stavros11** [comment]
> Thank you @stavros11, very good. Could you please try to clone directly inside `/tmp/`?

I don't think this is possible because the qrccluster virtual machines do not have git and I am not sure how to access the /tmp/ of the virtual machine (for example reem2) from saadiyat. 

Also in the CI we do not clone qibolab, I think it is automatically cloned to saadiyat when the workflow is launched and then we assume that we are inside the repository directory so we do `pip install .`. I tried copying all files from . to /tmp/ and doing `pip install .` there but the coverage hangs like before.

**Comment by scarrazza** [comment]
> I don't think this is possible because the qrccluster virtual machines do not have git and I am not sure how to access the /tmp/ of the virtual machine (for example reem2) from saadiyat.

I have just installed git in all VMs.

> I tried copying all files from . to /tmp/ and doing pip install . there but the coverage hangs like before.

Strange, that should work. I would suggest to give a try by cloning and executing pytest manually in one of VM's `/tmp/` folder (now you can access git) and check the coverage values, if no changes are visible then the symlink is working fine, otherwise we have to investigate why copy is not working and, maybe, try rsync instead.

**Comment by stavros11** [comment]
> I would suggest to give a try by cloning and executing pytest manually in one of VM's `/tmp/` folder (now you can access git) and check the coverage values, 

If I manually ssh to the virtual machine as admin and do that the coverage is higher than what is uploaded here. For example in test_instruments_qblox only 3 lines are missing, instead of 88 in the uploaded report. So it seems that the CI coverage misses the symlinked test files.

> otherwise we have to investigate why copy is not working and, maybe, try rsync instead.

I think the issue is not with copy/clone but with `pip install .`. Maybe because the conda environment is in NFS, pip installation hangs if it runs as part of a slurm script, either with srun/sbatch or if I open an srun bash in the node.

**Comment by scarrazza** [comment]
Right, thanks for checking. You could try to install dependencies with pip and them run pytest locally, eventually with https://docs.pytest.org/en/7.2.x/how-to/usage.html#calling-pytest-through-python-m-pytest

**Comment by scarrazza** [comment]
Following the discussion today, I would suggest to hold this implementation until we have a stable lab configuration. @AleCandido any thoughts?

**Comment by AleCandido** [comment]
As we discussed during the meeting: I have the feeling that there won't be a stable lab in a while, if not ever (in the end, the goal of the lab is trying to scale as much as possible), unless there will be a dedicated fridge for stability.
In any case, it won't happen soon.

What I would suggest is to keep the workflow and merge it, but changing the event just to:
```
on:
    workflow_dispatch:
```
It is more or less the same of the current label trick, but it will allow us to trigger it both manually (with a button in the Actions tab, the equivalent of the label trick - just loosing the check association to the PR) and automatically (through the GitHub API).

In the end, it is useful to have a bunch of checks and even to measure the coverage, but definitely we can't update the report to Codecov, it would be misleading.
In order to make the report comfortably accessible we can upload both the `htmlcov` folder and the `coverage.xml` file as an artifact, in case we want to inspect the test coverage, just using [`actinons/upload-artifact`](https://github.com/actions/upload-artifact).

File: .github/workflows/selfhosted.yml
Comment (AleCandido) [inline .github/workflows/selfhosted.yml:74]: Since you're just replacing `$platform`, the script is better committed to the repo, and then you can `sed` replace the placeholder (maybe call it `$PLATFORM` to make it more evident).

Comment (AleCandido) [reply .github/workflows/selfhosted.yml:74]: Git preserves the file mode, so even the `chmod +x` can be commited.

File: .github/workflows/selfhosted.yml
Comment (AleCandido) [inline .github/workflows/selfhosted.yml:75]:
```suggestion
        sed -i 's/$PLATFORM/'${{ matrix.platform }}'/' my_slurm.bash
        srun -p ${{ matrix.platform }} my_slurm.bash
```

(the `sed` command has to be tested and possibly debugged)

**Comment by scarrazza** [comment]
@stavros11 could you please firm that this PR requires fixes in order to obtain the expected coverage when running inside the `/tmp/` folder?

File: .github/workflows/selfhosted.yml
Comment (stavros11) [reply .github/workflows/selfhosted.yml]: Many thanks for this suggestion, it works and is much much cleaner than what I was doing. The only thing is that I placed the script in the rootdir so that I do not need to use longer dirs in the workflow, but if you think it is better somewhere else (eg. within the .github folder, we can move it).

**Comment by stavros11** [comment]
> @stavros11 could you please firm that this PR requires fixes in order to obtain the expected coverage when running inside the `/tmp/` folder?

After the latesh push (and merging main), it works: it generates the coverage.xml in the /tmp/ but then I copy it back to the original qibolab directory on saadiyat. I think from there we can upload it wherever we decide.

Currently I am using `cat` to prove that coverage.xml is indeed generated (see [here](https://github.com/qiboteam/qibolab/actions/runs/4204557194/jobs/7295392857)) but this is not readable. Ideally we would like to upload it somewhere. I agree with @AleCandido that uploading to Codecov would be misleading because then the coverage will depend on what labels we use each time and at this stage it is not productive to test all different QPUs for every small change. Maybe we could upload it as artifact.

Two othe smaller issues are:
* I am not sure why codecov complains given that I have not changed anything related to the workflow running on GitHub or the code in this PR.
* There is a `PytestUnknownMarkWarning: Unknown pytest.mark.qpu` warning, however despite this pytest seems to still recognize the marker and run the proper tests on slurm.

File: .github/workflows/selfhosted.yml
Comment (AleCandido) [inline .github/workflows/selfhosted.yml:65]:
```suggestion
- uses: actions/upload-artifact@v3
  with:
    name: coverage-from-self-hosted
    path: coverage.xml
```

Comment (AleCandido) [inline .github/workflows/selfhosted.yml:65]:
```suggestion
- uses: actions/upload-artifact@v3
  with:
    name: coverage-from-self-hosted
    path: coverage.xml
```

Comment (AleCandido) [reply .github/workflows/selfhosted.yml:65]: In particular, this is to upload the file itself, such that is easier to download and manipulate.

To make something readable, I would simply generate the html as well, and upload that one. Let me have a go at it.

**Comment by stavros11** [comment]
> What I would suggest is to keep the workflow and merge it, but changing the event just to:
> 
> ```
> on:
>     workflow_dispatch:
> ```
> 
> It is more or less the same of the current label trick, but it will allow us to trigger it both manually (with a button in the Actions tab, the equivalent of the label trick - just loosing the check association to the PR) and automatically (through the GitHub API).

This is fine for me, the only issue is whether it allows us to pass the platform as an argument, but I have not checked the docs so maybe it is possible. Currently we are using `run-on-*` labels and `*` specifies the platform - it is not ideal because you have to remove and put back the label every time but it still works. Maybe we could keep both.

**Comment by AleCandido** [comment]
Just rebased on `main`, to have the updated branch.

File: .github/workflows/selfhosted.yml
Comment (stavros11) [inline .github/workflows/selfhosted.yml:66]: I think you also need to copy the htmlcov folder from /tmp/ to qibolab rootdir, in the qputests.sh script.

**Comment by AleCandido** [comment]
> This is fine for me, the only issue is whether it allows us to pass the platform as an argument, but I have not checked the docs so maybe it is possible. Currently we are using `run-on-*` labels and `*` specifies the platform - it is not ideal because you have to remove and put back the label every time but it still works. Maybe we could keep both.

I see the problem, let me investigate if there is a suitable solution available. I'm pretty sure you can trigger parameterized workflows from the API (i.e. you can pass parameter values if you call it using `curl` or another http client).
A simple replacement would be to have a ChatOps bot, e.g. done with [Probot](https://probot.github.io/). In practice, it will be the exact same of labels, but more flexible, if we ever need something else.

File: .github/workflows/selfhosted.yml
Comment (AleCandido) [reply .github/workflows/selfhosted.yml:66]: You're definitely right, implemented in 1447f5b

**Comment by AleCandido** [comment]
@stavros11 now I'm rather confused: for `coverage.xml` everything is working smoothly, and locally I can generate the `htmlcov` folder.
Now the `mv` is in principle also in the dispatched script, but the workflow is not able to find it afterward.

I added a couple of `ls` to attempt to understand, do you have any better idea?

**Comment by AleCandido** [comment]
> @stavros11 now I'm rather confused: for `coverage.xml` everything is working smoothly, and locally I can generate the `htmlcov` folder. Now the `mv` is in principle also in the dispatched script, but the workflow is not able to find it afterward.
> 
> I added a couple of `ls` to attempt to understand, do you have any better idea?

I take it back: I didn't carefully read the sbatched script, my bad.

**Comment by stavros11** [comment]
> I take it back: I didn't carefully read the sbatched script, my bad.

Yes, I think it was missing the pyproject.toml or you needed to pass the coverage options in the script.

**Comment by AleCandido** [comment]
> Yes, I think it was missing the pyproject.toml or you needed to pass the coverage options in the script.

I actually chose the first option, such that it will be kept consistent with future changes as well :)

**Comment by AleCandido** [comment]
Ok, now we have the first report from the QPU: https://github.com/qiboteam/qibolab/suites/11049780903/artifacts/561502413

Goal achieved: we can download a readable report, if ever needed.
Goal missed: you need a few operations, rather than just opening a web page (but just a few: download the file, unzip, double-click on `htmlcov/index.html`).

I'd say is enough for the time being. If ever we will start looking continuously at these reports, we can always try to engineer a better solution :)

**Review by AleCandido — APPROVED** [review]

**Comment by scarrazza** [comment]
Good point, so instead of uploading to codecov and messing the default coverage without QPU, the report becomes an artifact. I would suggest to add [this step](https://github.com/NNPDF/nnpdf/blob/bea644cfb02086d6f8ae2ffe9f660a6f9d5f5d0e/.github/workflows/fitbot.yml#L114) to the self-hosted the workflow. Namely, a final step in the workflow which writes a comment in the PR with the link to the report.

**Comment by stavros11** [comment]
> I'd say is enough for the time being. If ever we will start looking continuously at these reports, we can always try to engineer a better solution :)

Looks good to me too, thanks. One last point is that we are using the `-m qpu` option which skips all tests that are not marked for qpu, so the coverage is lower than the github workflow. On qpu we can run all tests if we want. If we do this we should get higher coverage than codecov, however it will also take slightly longer time on the qpu queue and potentially blocking other jobs.

> I would suggest to add [this step](https://github.com/NNPDF/nnpdf/blob/bea644cfb02086d6f8ae2ffe9f660a6f9d5f5d0e/.github/workflows/fitbot.yml#L114) to the self-hosted the workflow. Namely, a final step in the workflow which writes a comment in the PR with the link to the report.

That's also a good idea, @AleCandido let me know if you will add it, otherwise I can have a look.

**Comment by AleCandido** [comment]
> That's also a good idea, @AleCandido let me know if you will add it, otherwise I can have a look. If it is not very hard we could even have the bot print the coverage percentage (read from the xml) in the comment.

I should be able to do it later on today (or at most tomorrow). In case I forget, remind me :)

**Comment by scarrazza** [comment]
Looks good, including the :atom: for qubit! If you could add the platform name in "Run on QPU {{ $matrix.platform }}" would be even better!

**Comment by AleCandido** [comment]
> Looks good, including the :atom: for qubit! If you could add the platform name in "Run on QPU {{ $matrix.platform }}" would be even better!

Actually the "correct" one should be :atom_symbol: (since it is the atom emoji), while :atom: is the symbol of the Atom editor (now dead...).
However, the chosen one looked nicer :)

**Comment by scarrazza** [comment]
Indeed, that is perfect.

**Comment by AleCandido** [comment]
Anything missing?

**Comment by stavros11** [comment]
From my side, this is ready to merge.

**Review by scarrazza — APPROVED** [review]
I agree, thank you @stavros11 and @AleCandido.

**Pull request merged**
